# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/s0m-a/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

**I'm choosing Lane 1**: Ranking Signal Analysis.

**My question is**: which observable content and search signals are associated with higher visibility (impressions) and clicks?

**Reason**: As a software developer trying to build an online presence, I've experienced firsthand how crowded the web has become. Understanding which signals correlate with visibility: word count, content age, position, CTR, engagement; would tell me what actually matters when creating content, not just what feels right.

In [62]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

<h1>DECISION</h1>
    Which content signals should a creator or SEO strategist prioritize when   deciding how to improve a page's search visibility and click rate?

<h1>ACTION</h1>
   A content editor or developer uses the signal report to decide: 
       Should I write longer content? 
       Should I target lower-competition queries?            
       Should I focus on improving my page-1 CTR?
       should I recover a declining page first?
<h1>COST OF A WRONG CALL</h1>
    False insight (wrong signal): A developer invests weeks making content longer when word count isn't actually correlated with impressions → wasted effort on the wrong lever.


=====================================================================================
    
   Missed signal: Failing to spot that page-1 positions with low CTR are the highest-value opportunity → leaves clicks on the table when a simple title rewrite could help.

====================================================================================
    
   Overclaiming causation: Saying "longer articles rank better" when the data only shows correlation -> misleads the action

In [65]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

Looking at the raw data, impression_tier has four levels: low, moderate, good, excellent. Most pages appear to cluster in the lower tiers, this is the crowding problem I want to quantify. I'll use this column and position_tier to see which signal combinations appear in high-visibility pages versus low-visibility ones.

In [68]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv('../../data/raw/content_refresh_anonymized.csv')

#====1.How crowded is visibility? 
tier_order = ['low', 'moderate', 'good', 'excellent']
tier_count = df['impression_tier'].value_counts()
tier_pct = (tier_count / len(df) * 100).round(1)
print("Impression tier distbution")
result = pd.DataFrame({'count': tier_count, 'pct': tier_pct})
print(result)
#====2. Median CTR by position tier (filter out avg_position == 0)
has_pos = df[df['avg_position'] > 0]
result = has_pos.groupby('position_tier')['ctr'].median().sort_values(ascending=False)
print("Median CTR by position tier:")
print(result)


#====3. How many pages are 'good' impression tier but still declining?
goodAndDeclining = df[(df['impression_tier'] == 'good') & (df['trend_direction'] == 'down')]

print(f"\nPages with 'good' impressions that are STILL declining: {len(goodAndDeclining):,}")
print(f"That's {len(goodAndDeclining)/len(df[df['impression_tier']=='good']):.1%} of all 'good' impression pages")

Impression tier distbution
                 count   pct
impression_tier             
low              11248  37.5
moderate         10469  34.9
good              7205  24.0
excellent         1078   3.6
Median CTR by position tier:
position_tier
page_1      0.16
striking    0.11
page_3_5    0.03
deep        0.00
top_3       0.00
Name: ctr, dtype: float64

Pages with 'good' impressions that are STILL declining: 4,223
That's 58.6% of all 'good' impression pages


<h2>Three numbers from the starter dataset support studying ranking signals:</h2>

1. Visibility is highly concentrated :  72.4% of the 30,000 pages in this dataset sit at low or moderate impression tiers, meaning most content gets little to no search exposure. Only 3.6% reach excellent. The crowding problem is real and measurable.
2. Position tier is the strongest observable CTR signal: median CTR at page_1 (0.16%) is roughly 5× higher than at page_3_5 (0.03%), and near zero for deeper positions. This is not surprising, but quantifying it establishes position as the dominant signal to account for in any analysis.
3. High visibility does not guarantee stability: 58.6% of pages with 'good' impressions are still declining in trend. This means reaching visibility is necessary but not sufficient.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [71]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


<h1>What this analysis can say:</h1>

We observed that 72.4% of pages in this dataset have low or moderate search visibility

Position tier is associated with higher CTR; pages at page_1 have a median CTR roughly 5× that of page_3_5 pages in this dataset

Among pages with good impressions, 58.6% show a declining trend; suggesting that signals beyond volume matter for stability

These findings are directional and observational; useful for prioritising content decisions, not for making guarantees

<h1>What this analysis cannot say: </h1>
    
Improving position causes higher CTR:  we observe correlation; we ran no experiment
    
These are Google's ranking factors: we observe search outcomes, not the algorithm

Fixing these signals will reverse a page's decline

Anything identifying a specific client, URL, or query; all identifiers are demo

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.